### PR TITLE
enhancement: Migrate to C++17

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature:** This PR migrates the project over to C++17, which should be a rather painless process. ~~Famous last words.~~

## Feature Details
In the past, compiler support has been a major concern when choosing C++ versions.

According to the [official compiler support documents](https://en.cppreference.com/w/cpp/compiler_support/17), all but the Intel, IBM and Sun/Oracle compilers have full support for C++17 core features. Two of those compilers don't even have full C++11 support yet, so I don't think that's something to worry about.

The [oldest version of Debian under official LTS](https://en.wikipedia.org/wiki/Debian_version_history#Release_table) (Buster) ships [GCC 8.3](https://packages.debian.org/buster/gcc) and [Clang 7](https://packages.debian.org/buster/clang), both providing full support for C++17. The [oldest actively maintained* version of Visual Studio](https://en.wikipedia.org/wiki/Visual_Studio#History) (2017) should also have full support for C++17 language features.

Library support is more mixed, although the vast majority of features are supported on all these platforms. (Also, I'm not sure how much of that stuff we are actually going to use.)

_*The 2013 and 2015 versions are still technically under LTS, but received no updates in the past 7 years._

**If you want to test this PR, please note that you might have to delete the `build` and/or `vcpkg` directories to avoid reusing cached builds.**

## Usage Examples
There are some TODO items for newer C++ versions. I'm sure we can find plenty of uses for them elsewhere as well, once we are more familiar with them.

Also, replacing `::type` and `::value` with `_t` / `_v` functions is not technically necessary, but quyykk requested them. I guess that's a usecase?

## Testing Done
The game runs. I checked with `__cplusplus` to make sure the game was compiled with C++17.

### Automated Tests Added
N/A

## Performance Impact
N/A. If anything, this should be a step towards improving performance.

### CODE REVIEW

File: source/WeightedList.h [hunk 1]
```diff
@@ -48,9 +48,9 @@ class WeightedList {
 	// Average the result of the given function by the choices' weights.
 	template <class Callable>
 	typename std::enable_if<
-		std::is_arithmetic<typename std::result_of<Callable&&(const Type&&)>::type>::value,
+		std::is_arithmetic<typename std::invoke_result<Callable&&, const Type&&>::type>::value,
 		// The return type of WeightedList::Average, if the above test passes:
-		typename std::result_of<Callable&&(const Type&&)>::type
+		typename std::invoke_result<Callable&&, const Type&&>::type
```
Comment (quyykk) [inline source/WeightedList.h:53]: Can you switch to using the `_t`/`_v` versions while you're at it? So like `typename std::enable_if<...>::type` => `std::enable_if_t<...>`. In other places too

File: tests/unit/src/comparators/test_byGivenOrder.cpp [hunk 2]
```diff
@@ -66,7 +66,8 @@ SCENARIO( "Test basic ByGivenOrder functionality." , "[ByGivenOrder]" ) {
 			std::vector<int> toSort = { 2, 4, 6, 8, 5, 1, 3 };
 			const std::vector<int> expectedOrder = { 4, 2, 8, 6, 1, 3, 5 };
 
-			std::random_shuffle(toSort.begin(), toSort.end());
+			unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+			std::shuffle(toSort.begin(), toSort.end(), std::default_random_engine(seed));
```
Comment (quyykk) [inline tests/unit/src/comparators/test_byGivenOrder.cpp:70]: Might be better to use `random_device` + `mt` here, like in the sample code here: https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution

File: source/WeightedList.h (hunk 1 above)
Comment (tibetiroka) [reply source/WeightedList.h:53]: ~~Uhh... how do I do that?~~

Comment (quyykk) [reply source/WeightedList.h:53]: Remove `typename` and `::type` and add a `_t` to the name. For values remove `::value` and add `_v`

Comment (tibetiroka) [reply source/WeightedList.h]: I _might_ have also replaced every other ::value and ::type as well.

File: tests/unit/src/comparators/test_byGivenOrder.cpp [hunk 3]
```diff
@@ -66,7 +66,9 @@ SCENARIO( "Test basic ByGivenOrder functionality." , "[ByGivenOrder]" ) {
 			std::vector<int> toSort = { 2, 4, 6, 8, 5, 1, 3 };
 			const std::vector<int> expectedOrder = { 4, 2, 8, 6, 1, 3, 5 };
 
-			std::random_shuffle(toSort.begin(), toSort.end());
+			std::random_device rd;  // a seed source for the random number engine
+			std::mt19937 gen(rd()); // mersenne_twister_engine seeded with rd()
```
Comment (quyykk) [inline tests/unit/src/comparators/test_byGivenOrder.cpp:70]: Those comments seem unnecessary, they just repeat what the code already does.

**Review by quyykk — APPROVED** [review]

**Review by RisingLeaf — APPROVED** [review]
Not my area of code, but I would like to have access to some features of c++17 especially filesystem::remove 😀

**Comment by bene-dictator** [comment]
Could also advise affected people to clear their build cache.

**Pull request merged**